File: android/features/src/main/kotlin/dev/example/moodlenative/features/LmsForumPresentationSupport.kt
package dev.example.moodlenative.features

import dev.example.moodlenative.core.LmsHTMLTextFormatter
import dev.example.moodlenative.core.LmsResourceTreeFile
import dev.example.moodlenative.core.LmsResourceTreeNode
import dev.example.moodlenative.core.lmsPathComponents
import dev.example.moodlenative.core.singleLineDisplayTextOrNull
import dev.example.moodlenative.core.treeNodes
import java.net.URI
import java.time.Instant

data class LmsForumPostThreadNode(
    val post: LmsForumPost,
    val children: List<LmsForumPostThreadNode>,
) {
    val id: Int
        get() = post.id
}

data class LmsForumDiscussionPresentation(
    val id: Int,
    val title: String,
    val authorName: String,
    val modifiedAt: Instant?,
    val replyCount: Int,
    val unreadCount: Int,
    val previewText: String?,
)

data class LmsForumPostPresentation(
    val id: Int,
    val subject: String?,
    val bodyText: String?,
    val authorName: String,
    val isUnread: Boolean,
    val modifiedAt: Instant?,
    val attachmentCount: Int,
)

data class LmsForumPostThreadPresentationNode(
    val post: LmsForumPostPresentation,
    val children: List<LmsForumPostThreadPresentationNode>,
) {
    val id: Int
        get() = post.id
}

val LmsForumDiscussion.presentation: LmsForumDiscussionPresentation
    get() = LmsForumDiscussionPresentation(
        id = discussionID,
        title = displayTitle,
        authorName = authorName,
        modifiedAt = modifiedAt,
        replyCount = replyCount,
        unreadCount = unreadCount,
        previewText = previewText,
    )

val LmsForumDiscussion.displayTitle: String
    get() = when {
        subject.isNotEmpty() -> subject
        name.isNotEmpty() -> name
        else -> "投稿"
    }

val LmsForumDiscussion.previewText: String?
    get() = LmsHTMLTextFormatter.plainText(message).singleLineDisplayTextOrNull()

val LmsForumDiscussion.authorName: String
    get() = userFullName.trim().ifEmpty { "投稿者不明" }

val LmsForumDiscussion.modifiedAt: Instant?
    get() = modifiedTimestamp.takeIf { it > 0 }?.let { Instant.ofEpochSecond(it.toLong()) }

val LmsForumDiscussion.replyCount: Int
    get() = maxOf(0, numberOfReplies)

val LmsForumDiscussion.unreadCount: Int
    get() = maxOf(0, numberOfUnreadPosts)

val LmsForumPost.createdAt: Instant?
    get() = timeCreated.takeIf { it > 0 }?.let { Instant.ofEpochSecond(it.toLong()) }

val LmsForumPost.displaySubject: String?
    get() = subject.trim().ifEmpty { null }

val LmsForumPost.bodyText: String?
    get() = LmsHTMLTextFormatter.plainText(message)

val LmsForumPost.authorName: String
    get() = author.fullName.trim().ifEmpty { "投稿者不明" }

val LmsForumPost.isUnread: Boolean
    get() = unread

val LmsForumPost.modifiedAt: Instant?
    get() = timeModified.takeIf { it > 0 }?.let { Instant.ofEpochSecond(it.toLong()) } ?: createdAt

val LmsForumPost.presentation: LmsForumPostPresentation
    get() = LmsForumPostPresentation(
        id = id,
        subject = displaySubject,
        bodyText = bodyText,
        authorName = authorName,
        isUnread = isUnread,
        modifiedAt = modifiedAt,
        attachmentCount = attachments.size,
    )

fun List<LmsForumPost>.threadedPosts(): List<LmsForumPostThreadNode> {
    val sortedPosts = sortedWith(compareBy<LmsForumPost> { it.timeCreated }.thenBy { it.id })
    val postsByID = sortedPosts.associateBy { it.id }
    val childrenByParent = mutableMapOf<Int, MutableList<LmsForumPost>>()
    val roots = mutableListOf<LmsForumPost>()

    for (post in sortedPosts) {
        val parentID = post.parentID
        if (parentID != null && postsByID[parentID] != null) {
            childrenByParent.getOrPut(parentID) { mutableListOf() } += post
        } else {
            roots += post
        }
    }

    fun makeNode(post: LmsForumPost): LmsForumPostThreadNode =
        LmsForumPostThreadNode(
            post = post,
            children = childrenByParent[post.id].orEmpty().map(::makeNode),
        )

    return roots.map(::makeNode)
}

fun List<LmsForumPost>.threadPresentation(): List<LmsForumPostThreadPresentationNode> =
    threadedPosts().map { node -> node.presentation }

private val LmsForumPostThreadNode.presentation: LmsForumPostThreadPresentationNode
    get() = LmsForumPostThreadPresentationNode(
        post = post.presentation,
        children = children.map { child -> child.presentation },
    )

val LmsForumPostAttachment.uri: URI?
    get() = fileURL?.let { runCatching { URI(it) }.getOrNull() }

val LmsForumPostAttachment.pathComponents: List<String>
    get() = (filePath ?: "").lmsPathComponents

val LmsForumPostAttachment.displayName: String
    get() = fileName?.trim()?.ifEmpty { null } ?: "添付ファイル"

val List<LmsForumPostAttachment>.resourceTreeNodes: List<LmsResourceTreeNode>
    get() = mapNotNull { attachment ->
        val uri = attachment.uri ?: return@mapNotNull null
        LmsResourceTreeFile(
            id = attachment.id,
            title = attachment.displayName,
            url = uri,
            pathComponents = attachment.pathComponents,
        )
    }.treeNodes()

val LmsForumPostURLs.discussURI: URI?
    get() = discuss?.let { runCatching { URI(it) }.getOrNull() }

val LmsForumPostAuthorURLs.profileImageURI: URI?
    get() = profileImage?.let { runCatching { URI(it) }.getOrNull() }

File: ios/Shared/Sources/Features/LmsForumPresentationSupport.swift
import Foundation
import MoodleNativeCore
import MoodleNativeNetworking

public struct LmsForumPostThreadNode: Identifiable {
    public let post: LmsWebServiceClient.ForumPost
    public let children: [LmsForumPostThreadNode]

    public init(post: LmsWebServiceClient.ForumPost, children: [LmsForumPostThreadNode]) {
        self.post = post
        self.children = children
    }

    public var id: Int {
        post.id
    }
}

extension LmsWebServiceClient.ForumDiscussion {
    public var displayTitle: String {
        if subject.isEmpty == false {
            return subject
        }

        if name.isEmpty == false {
            return name
        }

        return "投稿"
    }

    public var previewText: String? {
        LmsHTMLTextFormatter.plainText(from: message).singleLineDisplayText
    }

    public var authorName: String {
        let trimmed = userFullName.trimmingCharacters(in: .whitespacesAndNewlines)
        return trimmed.isEmpty ? "投稿者不明" : trimmed
    }

    public var modifiedAt: Date? {
        guard modifiedTimestamp > 0 else {
            return nil
        }

        return Date(timeIntervalSince1970: TimeInterval(modifiedTimestamp))
    }

    public var replyCount: Int {
        max(0, numberOfReplies)
    }

    public var unreadCount: Int {
        max(0, numberOfUnreadPosts)
    }
}
extension LmsWebServiceClient.ForumPost {
    public var createdAt: Date? {
        guard timeCreated > 0 else {
            return nil
        }

        return Date(timeIntervalSince1970: TimeInterval(timeCreated))
    }

    public var displaySubject: String? {
        let trimmed = subject.trimmingCharacters(in: .whitespacesAndNewlines)
        return trimmed.isEmpty ? nil : trimmed
    }

    public var bodyText: String? {
        LmsHTMLTextFormatter.plainText(from: message)
    }

    public var authorName: String {
        let trimmed = author.fullName.trimmingCharacters(in: .whitespacesAndNewlines)
        return trimmed.isEmpty ? "投稿者不明" : trimmed
    }

    public var isUnread: Bool {
        unread
    }

    public var modifiedAt: Date? {
        guard timeModified > 0 else {
            return createdAt
        }

        return Date(timeIntervalSince1970: TimeInterval(timeModified))
    }
}

extension Array where Element == LmsWebServiceClient.ForumPost {
    public func threadedPosts() -> [LmsForumPostThreadNode] {
        let sortedPosts = sorted { lhs, rhs in
            if lhs.timeCreated == rhs.timeCreated {
                return lhs.id < rhs.id
            }

            return lhs.timeCreated < rhs.timeCreated
        }

        let postsByID = Dictionary(uniqueKeysWithValues: sortedPosts.map { ($0.id, $0) })
        var childrenByParent: [Int: [LmsWebServiceClient.ForumPost]] = [:]
        var roots: [LmsWebServiceClient.ForumPost] = []

        for post in sortedPosts {
            if let parentID = post.parentID, postsByID[parentID] != nil {
                childrenByParent[parentID, default: []].append(post)
            } else {
                roots.append(post)
            }
        }

        func makeNode(for post: LmsWebServiceClient.ForumPost) -> LmsForumPostThreadNode {
            LmsForumPostThreadNode(
                post: post,
                children: (childrenByParent[post.id] ?? []).map(makeNode)
            )
        }

        return roots.map(makeNode)
    }
}

extension LmsWebServiceClient.ForumPostAttachment {
    public var url: URL? {
        fileURL.flatMap(URL.init(string:))
    }

    public var pathComponents: [String] {
        (filePath ?? "").lmsPathComponents
    }

    public var displayName: String {
        let trimmed = fileName?.trimmingCharacters(in: .whitespacesAndNewlines) ?? ""
        return trimmed.isEmpty ? "添付ファイル" : trimmed
    }
}

extension Array where Element == LmsWebServiceClient.ForumPostAttachment {
    public var resourceTreeNodes: [LmsResourceTreeNode] {
        compactMap { attachment in
            guard let url = attachment.url else {
                return nil
            }

            return LmsResourceTreeFile(
                id: attachment.id,
                title: attachment.displayName,
                url: url,
                pathComponents: attachment.pathComponents
            )
        }
        .treeNodes()
    }
}

extension LmsWebServiceClient.ForumPostURLs {
    public var discussURL: URL? {
        discuss.flatMap(URL.init(string:))
    }
}

extension LmsWebServiceClient.ForumPostAuthorURLs {
    public var profileImageURL: URL? {
        profileImage.flatMap(URL.init(string:))
    }
}
